compiler/ast.nim

```diff
@@ -1472,22 +1473,26 @@ proc hasNilSon*(n: PNode): bool =
       return true
   result = false
 
-proc containsNode*(n: PNode, kinds: TNodeKinds): bool =
+proc containsNode*(n: PNode, predicate: proc(cn: PNode):bool): bool =
   if n == nil: return
   case n.kind
-  of nkEmpty..nkNilLit: result = n.kind in kinds
+  of nkEmpty..nkNilLit: result = predicate(n)
   else:
+    if predicate(n): return true
     for i in countup(0, sonsLen(n) - 1):
-      if n.kind in kinds or containsNode(n.sons[i], kinds): return true
+      if containsNode(n.sons[i], predicate): return true
+
+proc containsNode*(n: PNode, kinds: TNodeKinds): bool =
+  result = n.containsNode(proc(cn: PNode): bool =
+    cn.kind in kinds)
 
 proc hasSubnodeWith*(n: PNode, kind: TNodeKind): bool =
-  case n.kind
-  of nkEmpty..nkNilLit: result = n.kind == kind
-  else:
-    for i in countup(0, sonsLen(n) - 1):
-      if (n.sons[i].kind == kind) or hasSubnodeWith(n.sons[i], kind):
-        return true
-    result = false
+  result = n.containsNode(proc(cn: PNode): bool =
+    cn.kind == kind)
+
+proc hasSubnodeWith*(n: PNode, magic: TMagic): bool =
```

Comment:
You never use this and hasSubnodeWith shouldn't call a callback for every node, it's supposed to be very fast.